be/src/exec/workgroup/work_group.cpp

```diff
@@ -112,19 +120,75 @@ void WorkGroupManager::destroy() {
     _driver_worker_owner_manager.reset(nullptr);
     _scan_worker_owner_manager.reset(nullptr);
     _workgroups.clear();
+    update_metrics();
 }
 
 WorkGroupPtr WorkGroupManager::add_workgroup(const WorkGroupPtr& wg) {
     std::unique_lock write_lock(_mutex);
     auto unique_id = wg->unique_id();
     create_workgroup_unlocked(wg);
+    add_metrics(wg);
     if (_workgroup_versions.count(wg->id()) && _workgroup_versions[wg->id()] == wg->version()) {
         return _workgroups[unique_id];
     } else {
         return get_default_workgroup();
     }
 }
 
+void WorkGroupManager::add_metrics(const WorkGroupPtr& wg) {
+    std::call_once(_init_metrics, []() {
+        StarRocksMetrics::instance()->metrics()->register_hook("work_group_metrics_hook",
+                                                               [] { WorkGroupManager::instance()->update_metrics(); });
+    });
+
+    if (_wg_metrics.count(wg->name()) == 0) {
```

Comment:
lines 144 ~ 170, also should be put into call_once lambda.

A good code style is to define a private _add_metrics function, then invoke it in add_metrics:

```cpp
void WorkGroupManager::_add_metrics(const WorkGroupPtr& wg) {
       StarRocksMetrics::instance()->metrics()->register_hook("work_group_metrics_hook",
                                                               [] { WorkGroupManager::instance()->update_metrics(); });
      if (_wg_metrics.count(wg->name()) == 0) {
                    //cpu limit
        std::unique_ptr<starrocks::DoubleGauge> resource_group_cpu_limit_ratio(new DoubleGauge(MetricUnit::PERCENT));
        //cpu concurrent
        std::unique_ptr<starrocks::DoubleGauge> resource_group_cpu_use_ratio(new DoubleGauge(MetricUnit::PERCENT));
        //mem limit
        std::unique_ptr<starrocks::IntGauge> resource_group_mem_limit_bytes(new IntGauge(MetricUnit::BYTES));
        //mem concurrent
        std::unique_ptr<starrocks::IntGauge> resource_group_mem_allocated_bytes(new IntGauge(MetricUnit::BYTES));

        StarRocksMetrics::instance()->metrics()->register_metric("resource_group_cpu_limit_ratio",
                                                                 MetricLabels().add("name", wg->name()),
                                                                 resource_group_cpu_limit_ratio.get());
        StarRocksMetrics::instance()->metrics()->register_metric("resource_group_cpu_use_ratio",
                                                                 MetricLabels().add("name", wg->name()),
                                                                 resource_group_cpu_use_ratio.get());
        StarRocksMetrics::instance()->metrics()->register_metric("resource_group_mem_limit_bytes",
                                                                 MetricLabels().add("name", wg->name()),
                                                                 resource_group_mem_limit_bytes.get());
        StarRocksMetrics::instance()->metrics()->register_metric("resource_group_mem_allocated_bytes",
                                                                 MetricLabels().add("name", wg->name()),
                                                                 resource_group_mem_allocated_bytes.get());

        _wg_cpu_limit_metrics.emplace(wg->name(), std::move(resource_group_cpu_limit_ratio));
        _wg_cpu_metrics.emplace(wg->name(), std::move(resource_group_cpu_use_ratio));
        _wg_mem_limit_metrics.emplace(wg->name(), std::move(resource_group_mem_limit_bytes));
        _wg_mem_metrics.emplace(wg->name(), std::move(resource_group_mem_allocated_bytes));
    }
    _wg_metrics.emplace(wg->name(), wg->unique_id());
}

void WorkGroupManager::add_metrics(const WorkGroupPtr& wg) {
       std::call_once(_init_metrics, [&this, &wg]() {this->_add_metrics(wg);};
}

```

Reply:
we can't put add_metrics method into the call_once lambda, because add_metrics is a method of WorkGroupManager,  not WorkGroup, only hooks need call_once. And i will change the add_metrics method to private.

Reply:
lines 144 ~ 170 also need call-once,  concurrent access to add_metrics for the same workgroup is dangerous.

Reply:
add_metrics will only be called in add_workgroup

Reply:
in a very rare case,  add_workgroup may be invoked on two different workgroups with the same name. for an example, 
sysadmin alter the resource group and the delete and re-create a resource group with the same name.